File: baekjoon/DataStructure/005-10090-Counting Inversions-SegmentTree.py
import sys
from math import ceil, log2
input = sys.stdin.readline


def get_init_segment_tree(N):
    tree_height = ceil(log2(N))
    tree_size = (1 << (tree_height + 1))
    return [0] * tree_size


def find(tree, left, right, start, end, node):
    if start > right or end < left:
        return 0
    if left <= start and end <= right:
        return tree[node]
    
    mid = (start + end) // 2
    return find(tree,left,right,start,mid,node*2) + find(tree,left,right,mid+1,end,node*2+1)


def update(tree, val, diff, start, end, node):
    if start == end:
        tree[node] += diff
        return
    
    mid = (start + end) // 2
    if val <= mid: update(tree, val, diff, start, mid, node * 2)
    else: update(tree, val, diff, mid+1, end, node * 2 + 1)
    tree[node] = tree[node * 2] + tree[node * 2 + 1]


def main():
    N = int(input())
    data = list(map(int, input().split()))
    tree = get_init_segment_tree(N)
    ans = 0
    for i in data:
        # print(find(tree, i+1, size, 1, size, 1))
        ans += (i - 1) - find(tree, 1, i-1, 1, N, 1)
        update(tree, i, 1, 1, N, 1)

    print(ans)


if __name__ == '__main__':
    main()

File: baekjoon/DataStructure/005-10090-Counting Inversions-SegmentTree.cpp
#include <iostream>
#include <string>
#include <cmath>
#include <set>
#include <map>
#include <stack>
#include <queue>
#include <vector>
#include <algorithm>
#include <climits>
#define FASTIO cin.tie(NULL); cout.tie(NULL); ios::sync_with_stdio(false);
#define MAX 1000001
#define LL long long
#define INF 1e9

using namespace std;
int N;
LL MAP[MAX];
vector<int> SegTree;
LL answer = 0;

// Inversion Counting

void Init() {
	int Tree_Height = (int)ceil(log2(N));
	int Tree_Size = (1 << (Tree_Height + 1));
	SegTree.resize(Tree_Size);
}

void Update_SegTree(int Node, int S, int E, LL Val, LL Diff) {
	if (S == E) {
		SegTree[Node] += Diff;
		return;
	}
	int M = (S + E) / 2;
	if (Val <= M) {
		Update_SegTree(Node * 2, S, M, Val, Diff);
	}
	else {
		Update_SegTree(Node * 2 + 1, M + 1, E, Val, Diff);
	}
	SegTree[Node] = SegTree[Node * 2] + SegTree[Node * 2 + 1];
}

LL Find_Value(int Node, int S, int E, int Left, int Right) {
	if ((S > Right) || (Left > E)) {
		return 0;
	}
	if ((Left <= S) && (E <= Right)) {
		return SegTree[Node];
	}
	int M = (S + E) / 2;
	return Find_Value(Node * 2, S, M, Left, Right) + Find_Value(Node * 2 + 1, M + 1, E, Left, Right);
}

void Find_Answer() {
	cin >> N;
	Init();
	for (int i = 1; i <= N; i++) {
		int X;
		cin >> X;
		answer += (X - 1) - Find_Value(1, 1, N, 1, X - 1);
		Update_SegTree(1, 1, N, X, 1);
	}
	cout << answer << "\n";
}

int main() {
	FASTIO
	Find_Answer();

	return 0;
}